src/message.rs

```diff
@@ -214,9 +214,10 @@ impl Message {
             _ => return Ok(None)
         };
 
+        let mut body = self.body.clone();
```

Comment:
Better would probably to `RefCell<Option<>>` the results as a memo. There's no reason to demarshal a body more than once.